hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -830,8 +831,35 @@ def debug_data(self, data: Union[ConjectureData, ConjectureResult]) -> None:
             f"{', ' + data.output if data.output else ''}"
         )
 
+    def observe_for_provider(self) -> AbstractContextManager:
+        def on_observation(observation: Observation) -> None:
+            assert observation.type == "test_case"
+            # because lifetime == "test_function"
+            assert isinstance(self.provider, PrimitiveProvider)
+            # only fire if we actually used that provider to generate this observation
+            if not self._switch_to_hypothesis_provider:
+                self.provider.on_observation(observation)
+
+        # adding this callback enables observability. Use a nullcontext
+        # if the backend won't use observability.
+        return (
+            with_observation_callback(on_observation)
+            if (
+                self.settings.backend != "hypothesis"
+                # only for lifetime = "test_function" providers (guaranteed
+                # by this isinstance check)
+                and isinstance(self.provider, PrimitiveProvider)
+                # and the provider opted-in to observations
+                and self.provider.add_observability_callback
+            )
+            else nullcontext()
+        )
```

Comment:
Seems cleaner as an if-statement and two returns than a ternary expression and one.